language/move-bytecode-verifier/src/control_flow.rs

```diff
@@ -7,268 +7,82 @@
+        for node in back {
+            let node = partition.containing_loop(*node);
 
-        // Pop if last continue
-        match instr {
-            // Back jump/"continue"
-            Bytecode::Branch(target) | Bytecode::BrTrue(target) | Bytecode::BrFalse(target)
-                if is_back_edge(cur_instr, *target) =>
-            {
-                let (_cur_loop_head, last_continue) = loop_stack.last().unwrap();
-                if cur_instr == *last_continue {
-                    loop_stack.pop();
-                }
+            if node != head {
+                body.insert(node);
             }
-            _ => (),
         }
-    }
-    Ok(())
-}
 
-fn is_back_edge(cur_instr: CodeOffset, target_instr: CodeOffset) -> bool {
-    target_instr <= cur_instr
-}
-
-// All back jumps are only to the current loop
-fn check_continues(context: &ControlFlowVerifier, labels: &[Label]) -> PartialVMResult<()> {
-    check_code(context, labels, |loop_stack, cur_instr, instr| {
-        match instr {
-            // Back jump/"continue"
-            Bytecode::Branch(target) | Bytecode::BrTrue(target) | Bytecode::BrFalse(target)
-                if is_back_edge(cur_instr, *target) =>
-            {
-                let (cur_loop_head, _last_continue) = loop_stack.last().unwrap();
-                if target != cur_loop_head {
-                    // Invalid back jump. Cannot back jump outside of the current loop
-                    Err(context.error(StatusCode::INVALID_LOOP_CONTINUE, cur_instr))
-                } else {
-                    Ok(())
+        let mut frontier: Vec<_> = body.iter().cloned().collect();
+        while let Some(node) = frontier.pop() {
+            for pred in summary.pred_edges(node) {
+                let pred = partition.containing_loop(*pred);
+                if !summary.is_descendant(head, pred) {
+                    return err(StatusCode::INVALID_LOOP_SPLIT, summary.block(pred));
                 }
-            }
-            _ => Ok(()),
-        }
-    })
-}
 
-fn check_breaks(context: &ControlFlowVerifier, labels: &[Label]) -> PartialVMResult<()> {
-    check_code(context, labels, |loop_stack, cur_instr, instr| {
-        match instr {
-            // Forward jump/"break"
-            Bytecode::Branch(target) | Bytecode::BrTrue(target) | Bytecode::BrFalse(target)
-                if !is_back_edge(cur_instr, *target) =>
-            {
-                match loop_stack.last() {
-                    Some((_cur_loop_head, last_continue))
-                        if target > last_continue && *target != last_continue + 1 =>
-                    {
-                        // Invalid loop break. Must break immediately to the instruction after
-                        // the last continue
-                        Err(context.error(StatusCode::INVALID_LOOP_BREAK, cur_instr))
-                    }
-                    _ => Ok(()),
+                if pred != head && body.insert(pred) {
```

Comment:
minor nit, I am a simpleton that can never remember the bool orderings around insert. So maybe `let was_inserted = prev != head && bool.insert(pred)` would help me